File: polyglot/java/layer_size_calculator.java
package polyglot.java;

import java.io.BufferedReader;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Dockerfile Layer Size Calculator for shipcheck tool.
 * Parses a Dockerfile and calculates cumulative layer sizes with detailed breakdowns.
 */
public class LayerSizeCalculator {

    private static final long DEFAULT_RUN_SIZE = 1024 * 1024L; // 1MB per RUN as baseline
    private static final long DEFAULT_COPY_SIZE_ESTIMATE = 50 * 1024; // 50KB per COPY/ADD line
    private static final long DEFAULT_FROM_BASE_SIZE = 100 * 1024 * 1024L; // 100MB base image estimate

    public record LayerInfo(String instruction, String details, long size) {}

    public record Summary(long totalSize, List<LayerInfo> layers) {
        public double toMegabytes() {
            return (double) totalSize / (1024 * 1024);
        }
    }

    private static final Pattern LINE_PATTERN = Pattern.compile(
        "^\\s*(?:#\\s*|/\\*.*?\\*/\\s*)?(FROM|RUN|COPY|ADD|CMD|ENTRYPOINT|ENV|LABEL|ARG|WORKDIR|" +
        "EXPOSE|USER|VOLUME|HEALTHCHECK|SHELL|STOPSIGNAL|ONBUILD|MAINTAINER|USER|EXPOSE|" +
        "CACHED|COMMIT)\\s*(.*)$"
    );

    /**
     * Main entry point demonstrating the calculator with a sample Dockerfile.
     */
    public static void main(String[] args) throws IOException {
        // Sample Dockerfile content for demonstration
        String sampleDockerfile = 
            "# Example multi-stage build\n" +
            "FROM node:18-alpine AS builder\n" +
            "RUN apk add --no-cache yarn && yarn install\n" +
            "COPY . /app/src\n" +
            "WORKDIR /app/src\n" +
            "RUN npm run build\n" +
            "\n" +
            "# Production stage\n" +
            "FROM node:18-alpine AS production\n" +
            "WORKDIR /app\n" +
            "COPY --from=builder /app/dist/ ./dist/\n" +
            "COPY package*.json ./\n" +
            "RUN npm install --production\n" +
            "CMD [\"node\", \"dist/index.js\"]\n";

        LayerSizeCalculator calculator = new LayerSizeCalculator();
        
        System.out.println("=== ShipCheck: Dockerfile Layer Size Calculator ===\n");
        
        Summary result = calculator.calculate(sampleDockerfile);
        
        printSummary(result);
    }

    /**
     * Parses the Dockerfile and calculates layer sizes.
     */
    public Summary calculate(String dockerfileContent) {
        List<LayerInfo> layers = new ArrayList<>();
        long runningTotal = 0L;
        String currentBaseImage = "";

        for (int i = 0; i < dockerfileContent.length(); i++) {
            if (dockerfileContent.charAt(i) == '\n') {
                // Process complete line
                int start = Math.max(0, i - 1);
                while (i > 0 && Character.isWhitespace(dockerfileContent.charAt(i - 1))) {
                    i--;
                }
                
                String line = dockerfileContent.substring(start, i + 1).trim();
                if (line.isEmpty() || line.startsWith("#")) {
                    continue;
                }

                Matcher matcher = LINE_PATTERN.matcher(line);
                if (!matcher.matches()) {
                    // Unknown instruction - treat as RUN with estimate
                    runningTotal += DEFAULT_RUN_SIZE;
                    layers.add(new LayerInfo("UNKNOWN", line, DEFAULT_RUN_SIZE));
                    continue;
                }

                String instruction = matcher.group(1).toUpperCase();
                String args = matcher.group(2);

                long layerSize = calculateLayerSize(instruction, args, currentBaseImage);
                
                if (instruction.equals("FROM")) {
                    runningTotal = 0L; // Fresh start
                    currentBaseImage = args.trim();
                    layers.add(new LayerInfo("FROM", "New base image: " + args, layerSize));
                } else if (instruction.equals("RUN")) {
                    runningTotal += layerSize;
                    layers.add(new LayerInfo("RUN", args.substring(0, Math.min(60, args.length())), layerSize));
                } else if (instruction.equals("COPY") || instruction.equals("ADD")) {
                    runningTotal += layerSize;
                    layers.add(new LayerInfo(instruction, args.substring(0, 40) + "...", layerSize));
                } else if (instruction.equals("CMD") || instruction.equals("ENTRYPOINT")) {
                    // Minimal size impact
                    runningTotal += 1024L;
                    layers.add(new LayerInfo(instruction, args.substring(0, 30), 1024));
                } else if (instruction.equals("ENV") || instruction.equals("LABEL") || 
                           instruction.equals("ARG") || instruction.equals("WORKDIR")) {
                    // Negligible size impact
                    runningTotal += 512L;
                    layers.add(new LayerInfo(instruction, args.substring(0, 30), 512));
                } else if (instruction.equals("EXPOSE") || instruction.equals("USER") || 
                           instruction.equals("VOLUME")) {
                    // Very minimal impact
                    runningTotal += 256L;
                    layers.add(new LayerInfo(instruction, args.substring(0, 30), 256));
                } else if (instruction.equals("HEALTHCHECK")) {
                    runningTotal += 1024L;
                    layers.add(new LayerInfo(instruction, "Health check configured", 1024));
                } else if (instruction.equals("SHELL") || instruction.equals("ONBUILD")) {
                    runningTotal += 512L;
                    layers.add(new LayerInfo(instruction, args.substring(0, 30), 512));
                }

                i++; // Move past the newline we already processed
            }
        }

        return new Summary(runningTotal, layers);
    }

    /**
     * Calculates estimated size for a specific instruction.
     */
    private long calculateLayerSize(String instruction, String args, String currentBaseImage) {
        switch (instruction.toUpperCase()) {
            case "FROM":
                // Base images vary significantly - use estimate based on common ones
                if (args.contains(":alpine")) return 50 * 1024 * 1024L;   // ~50MB
                if (args.contains(":debian") || args.contains(":ubuntu")) return 80 * 1024 * 1024L;
                if (args.contains(":slim")) return 70 * 1024 * 1024L;
                return DEFAULT_FROM_BASE_SIZE; // Default ~100MB
            case "RUN":
                // Estimate based on command length and common patterns
                long runSize = DEFAULT_RUN_SIZE;
                
                if (args.contains("apk add") || args.contains("apt-get install")) {
                    runSize += 2 * 1024 * 1024L; // Package manager overhead
                } else if (args.contains("npm install") || args.contains("yarn add")) {
                    runSize += 5 * 1024 * 1024L; // Node packages
                } else if (args.contains("pip install") || args.contains("cargo build")) {
                    runSize += 3 * 1024 * 1024L; // Python/Rust dependencies
                } else if (args.contains("curl") && args.contains("|")) {
                    runSize += 10 * 1024 * 1024L; // Downloading files
                } else if (args.contains("wget") || args.contains("git clone")) {
                    runSize += 5 * 1024 * 1024L; // Git/HTTP downloads
                }
                
                return Math.max(runSize, 1024 * 1024L);

            case "COPY":
            case "ADD":
                // Estimate based on number of source files mentioned
                String[] parts = args.split("\\s+");
                long copySize = DEFAULT_COPY_SIZE_ESTIMATE;
                
                for (int i = 0; i < parts.length - 1 && i < 5; i++) {
                    if (!parts[i].startsWith("--")) {
                        // Count source files/directories
                        String[] sources = parts[i].split("/");
                        copySize += Math.max(1, sources.length) * 2048L;
                    }
                }
                
                return copySize;

            default:
                return 512L; // Default small estimate for other instructions
        }
    }

    /**
     * Prints a formatted summary of the analysis.
     */
    private void printSummary(Summary result) {
        System.out.println("Total Estimated Size: " + String.format("%.2f", result.toMegabytes()) + " MB");
        System.out.println("Layers Analyzed: " + result.layers.size());

        if (!result.layers.isEmpty()) {
            System.out.println("\n--- Layer Breakdown ---");
            
            // Group layers by instruction type for cleaner output
            Map<String, List<LayerInfo>> grouped = new HashMap<>();
            for (LayerInfo layer : result.layers) {
                String key = layer.instruction;
                grouped.computeIfAbsent(key, k -> new ArrayList<>()).add(layer);
            }

            // Print each group
            for (Map.Entry<String, List<LayerInfo>> entry : sortedGroups(grouped)) {
                System.out.println("\n" + entry.getKey() + ":");
                long groupTotal = 0;
                
                for (LayerInfo layer : entry.getValue()) {
                    String displaySize = formatBytes(layer.size);
                    System.out.printf("  %s: %s\n", 
                        truncate(layer.details, 45), displaySize);
                    groupTotal += layer.size;
                }
                
                System.out.println(String.format("  Subtotal: %s (%.1f%% of total)", 
                    formatBytes(groupTotal), (groupTotal / (double) result.totalSize * 100)));
            }

            // Top contributors analysis
            printTopContributors(result);
        }
    }

    private void printTopContributors(Summary result) {
        System.out.println("\n--- Top Size Contributors ---");
        
        List<LayerInfo> sorted = new ArrayList<>(result.layers);
        sorted.sort((a, b) -> Long.compare(b.size(), a.size()));
        
        int limit = Math.min(5, sorted.size());
        for (int i = 0; i < limit; i++) {
            LayerInfo layer = sorted.get(i);
            System.out.printf("%d. %s: %s\n", 
                i + 1, truncate(layer.details, 40), formatBytes(layer.size));
        }

        // Identify potential optimization opportunities
        printOptimizationSuggestions(result);
    }

    private void printOptimizationSuggestions(Summary result) {
        System.out.println("\n--- Optimization Suggestions ---");
        
        boolean hasMultiStage = false;
        long multiStageSavings = 0L;
        
        for (LayerInfo layer : result.layers) {
            if (layer.instruction.equals("FROM")) {
                hasMultiStage = true;
                // Estimate savings from multi-stage builds
                String base = layer.details.split(":")[0];
                if (!base.contains("alpine") && !base.contains("slim")) {
                    multiStageSavings += 30 * 1024 * 1024L; // Assume ~30MB savings per non-alpine base
                }
            } else if (layer.instruction.equals("RUN") && 
                       layer.details.toLowerCase().contains("npm install")) {
                System.out.println("- Consider using multi-stage builds to reduce node_modules");
                break;
            } else if (layer.instruction.equals("COPY") || layer.instruction.equals("ADD")) {
                // Check for large files being copied
                String[] parts = layer.details.split("\\s+");
                for (String part : parts) {
                    if (!part.startsWith("--") && !part.startsWith(".") && 
                        !part.contains("/dist/") && !part.contains("/build/")) {
                        System.out.println("- Review COPY/ADD sources: " + truncate(part, 30));
                    }
                }
            }
        }

        if (hasMultiStage) {
            System.out.printf("- Multi-stage build detected. Estimated savings: %s\n", 
                formatBytes(multiStageSavings));
        } else {
            System.out.println("- Consider multi-stage builds to reduce final image size");
        }
    }

    private static List<Map.Entry<String, List<LayerInfo>>> sortedGroups(
            Map<String, List<LayerInfo>> grouped) {
        return new ArrayList<>(grouped.entrySet()).stream()
                .sorted(Map.Entry.comparingByKey())
                .toList();
    }

    private static String truncate(String str, int maxLen) {
        if (str.length() <= maxLen) return str;
        return str.substring(0, maxLen - 3) + "...";
    }

    private static String formatBytes(long bytes) {
        if (bytes < 1024) return bytes + " B";
        if (bytes < 1024 * 1024) return String.format("%.1f KB", bytes / 1024.0);
        if (bytes < 1024 * 1024 * 1024) return String.format("%.1f MB", bytes / (1024.0 * 1024));
        return String.format("%.1f GB", bytes / (1024.0 * 1024 * 1024));
    }
}

File: polyglot/typescript/layer_size_calculator.ts
import * as fs from 'fs';
import * as path from 'path';

export interface LayerInfo {
  instruction: string;
  lineNumber: number;
  estimatedSizeMB: number;
  description: string;
}

export interface SizeSummary {
  totalLayers: number;
  totalSizeMB: number;
  baseImageMB: number;
  runtimeCommandsMB: number;
  fileCopiesMB: number;
  layers: LayerInfo[];
}

interface DockerfileState {
  currentLayerSize: number;
  baseImageSize: number;
  runtimeCommandsSize: number;
  fileCopiesSize: number;
  layers: LayerInfo[];
  fromLine: number | null;
  fromTag: string | null;
}

interface ImageSizeEstimates {
  [tag: string]: number; // tag -> size in MB
}

const DEFAULT_BASE_IMAGES: ImageSizeEstimates = {
  'alpine': 5.0,
  'debian': 120.0,
  'ubuntu': 77.0,
  'node': 140.0,
  'python': 85.0,
  'golang': 130.0,
  'rust': 90.0,
};

function parseDockerfile(content: string): { instructions: string[]; lines: string[] } {
  const lines = content.split('\n');
  const instructions: string[] = [];
  
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i].trim();
    if (line && !line.startsWith('#')) {
      // Remove inline comments and whitespace
      const cleanLine = line.split('//')[0].split('/*')[0].trim();
      if (cleanLine) {
        instructions.push({ raw: line, clean: cleanLine, index: i });
      }
    }
  }
  
  return { instructions, lines };
}

function estimateBaseImageSize(tag: string | null): number {
  if (!tag) return DEFAULT_BASE_IMAGES['alpine'] || 50.0;
  
  // Extract base image tag (handle multi-stage builds)
  const parts = tag.split(':');
  const imageRef = parts[0];
  const version = parts.length > 1 ? parts.slice(1).join(':') : 'latest';
  
  if (DEFAULT_BASE_IMAGES[imageRef]) {
    return DEFAULT_BASE_IMAGES[imageRef] * parseFloat(version) || 
           DEFAULT_BASE_IMAGES[imageRef];
  }
  
  // Default fallback for unknown images
  return 50.0;
}

function estimateCopySize(source: string, isAdd: boolean = false): number {
  if (!source) return 0;
  
  // Check if it's a glob pattern or directory
  const isGlob = source.includes('*');
  const isDir = path.isAbsolute(source) || 
                 (path.dirname(source).length > 1 && !isGlob);
  
  let estimatedSizeMB: number;
  
  if (isGlob) {
    // Estimate glob as average file size * potential count
    // A typical glob might match 10-50 files of ~10KB each
    estimatedSizeMB = isDir ? 2.0 : 0.5;
  } else if (isDir) {
    // Directory copy includes all contents recursively
    estimatedSizeMB = 1.0 + Math.random() * 5.0;
  } else {
    // Single file - estimate based on common sizes
    const ext = path.extname(source).toLowerCase();
    switch (ext) {
      case '.js':
      case '.ts':
      case '.jsx':
      case '.tsx':
        estimatedSizeMB = 0.1 + Math.random() * 2.0;
        break;
      case '.html':
      case '.css':
      case '.scss':
      case '.less':
        estimatedSizeMB = 0.05 + Math.random();
        break;
      case '.json':
        estimatedSizeMB = 0.01 + Math.random() * 0.1;
        break;
      case '.sh':
      case '.bash':
        estimatedSizeMB = 0.02 + Math.random() * 0.3;
        break;
      default:
        // Default file estimate
        estimatedSizeMB = 0.05 + Math.random();
    }
  }
  
  return isAdd ? estimatedSizeMB * 1.1 : estimatedSizeMB;
}

function estimateRunCommandSize(command: string): number {
  const lowerCmd = command.toLowerCase().trim();
  
  // Commands that typically produce larger layers (install packages, etc.)
  const largeCommands = [
    /apt-get install/,
    /apk add/,
    /yum install/,
    /dnf install/,
    /brew install/,
    /pip install/,
    /npm install/,
    /cargo install/,
    /go get/,
    /mvn install/,
    /gradle build/,
  ];
  
  // Commands that produce moderate layers (build artifacts, logs)
  const mediumCommands = [
    /gcc /,
    /g++ /,
    /cmake /,
    /make /,
    /cargo build/,
    /go build/,
    /npm run build/,
    /webpack /,
    /vite /,
  ];
  
  // Commands that produce small layers (env vars, simple scripts)
  const smallCommands = [
    /echo /,
    /mkdir /,
    /touch /,
    /chown /,
    /chmod /,
    /ln -s/,
    /cat >/,
    /tee /,
  ];
  
  let sizeMB: number;
  
  if (largeCommands.some(r => r.test(lowerCmd))) {
    // Package installation can vary widely
    const pkgCount = lowerCmd.match(/(install|add)\s+[^ ]+/g)?.length || 1;
    sizeMB = 2.0 + Math.random() * 5.0 * (pkgCount > 1 ? pkgCount : 1);
  } else if (mediumCommands.some(r => r.test(lowerCmd))) {
    // Build artifacts
    sizeMB = 1.0 + Math.random() * 3.0;
  } else if (smallCommands.some(r => r.test(lowerCmd))) {
    // Minimal impact
    sizeMB = 0.05 + Math.random() * 0.2;
  } else {
    // Unknown command - estimate conservatively
    const wordCount = lowerCmd.split(/\s+/).length;
    sizeMB = 0.1 + Math.min(2.0, wordCount * 0.05);
  }
  
  return Math.max(sizeMB, 0.05); // Minimum non-zero for any RUN command
}

function analyzeDockerfile(content: string): SizeSummary {
  const { instructions, lines } = parseDockerfile(content);
  
  let state: DockerfileState = {
    currentLayerSize: 0,
    baseImageSize: 0,
    runtimeCommandsSize: 0,
    fileCopiesSize: 0,
    layers: [],
    fromLine: null,
    fromTag: null,
  };
  
  for (const { raw, clean, index } of instructions) {
    const lineNum = lines[index].trim().length > 0 ? index + 1 : index;
    
    // Parse instruction and arguments
    const parts = clean.split(/\s+/);
    if (parts.length === 0 || !parts[0]) continue;
    
    const instruction = parts[0];
    const args = parts.slice(1).join(' ').trim();
    
    let layerSize: number = 0;
    let description: string = '';
    
    switch (instruction.toUpperCase()) {
      case 'FROM':
        state.fromLine = lineNum;
        state.fromTag = args || 'alpine';
        state.baseImageSize = estimateBaseImageSize(state.fromTag);
        layerSize = state.baseImageSize;
        description = `Base image: ${state.fromTag}`;
        break;
        
      case 'RUN':
        // RUN commands create layers from their output
        if (args) {
          const runSize = estimateRunCommandSize(args);
          layerSize = runSize;
          state.runtimeCommandsSize += runSize;
          description = `Runtime command: ${args.substring(0, 50)}${args.length > 50 ? '...' : ''}`;
        } else {
          // Empty RUN (rare but possible)
          layerSize = 0.01;
          state.runtimeCommandsSize += 0.01;
          description = `Empty RUN command`;
        }
        break;
        
      case 'COPY':
      case 'ADD':
        if (args) {
          // Handle multiple sources or source + destination
          const isAdd = instruction.toUpperCase() === 'ADD';
          
          // Split into sources and destination
          let dest: string | null = null;
          let sources: string[] = [];
          
          // Last argument is typically the destination (unless it's a glob pattern)
          if (!args.includes('*')) {
            const parts = args.split(/\s+/);
            dest = parts[parts.length - 1];
            sources = parts.slice(0, -1).join(' ');
          } else {
            // Glob patterns - last part is destination
            const globParts = args.split(/\s+/);
            if (globParts[globParts.length - 1].includes('*')) {
              dest = globParts[globParts.length - 1];
              sources = globParts.slice(0, -1).join(' ');
            } else {
              // Multiple files with same destination pattern
              sources = args;
              dest = null;
            }
          }
          
          let copySize: number = 0;
          if (sources) {
            const estimatedCopySize = estimateCopySize(sources, isAdd);
            layerSize = estimatedCopySize;
            state.fileCopiesSize += estimatedCopySize;
            description = `Copy/ADD: ${sources.substring(0, 40)}${sources.length > 40 ? '...' : ''}`;
          } else {
            // Edge case - no sources specified
            layerSize = 0.01;
            description = `COPY/ADD without source`;
          }
        } else {
          layerSize = 0.01;
          description = `Empty COPY/ADD command`;
        }
        break;
        
      case 'ENV':
      case 'ARG':
      case 'LABEL':
      case 'EXPOSE':
      case 'CMD':
      case 'ENTRYPOINT':
      case 'USER':
      case 'WORKDIR':
      case 'VOLUME':
      case 'HEALTHCHECK':
        // These create metadata layers but minimal size impact
        layerSize = 0.01;
        description = `${instruction}: ${args.substring(0, 40)}${args.length > 40 ? '...' : ''}`;
        break;
        
      case 'MAINTAINER':
      case 'STOPSIGNAL':
      case 'SHELL':
      case 'ONBUILD':
      case 'PLATFROM': // Typo in older Dockerfiles
      default:
        layerSize = 0.01;
        description = `${instruction}: ${args.substring(0, 40)}${args.length > 40 ? '...' : ''}`;
    }
    
    state.currentLayerSize += layerSize;
    state.layers.push({
      instruction: instruction,
      lineNumber: lineNum,
      estimatedSizeMB: Math.round(layerSize * 100) / 100,
      description,
    });
  }
  
  return {
    totalLayers: state.layers.length,
    totalSizeMB: Math.round(state.currentLayerSize * 100) / 100,
    baseImageMB: Math.round(state.baseImageSize * 100) / 100,
    runtimeCommandsMB: Math.round(state.runtimeCommandsSize * 100) / 100,
    fileCopiesMB: Math.round(state.fileCopiesSize * 100) / 100,
    layers: state.layers,
  };
}

export function analyzeDockerfileSync(dockerfilePath: string): SizeSummary {
  const content = fs.readFileSync(dockerfilePath, 'utf-8');
  return analyzeDockerfile(content);
}

export async function analyzeDockerfileAsync(dockerfilePath: string): Promise<SizeSummary> {
  const content = await fs.promises.readFile(dockerfilePath, 'utf-8');
  return analyzeDockerfile(content);
}

// Demo / Entry point
if (require.main === module) {
  // Default demo with a sample Dockerfile
  const sampleDockerfile = `FROM node:16-alpine AS builder
WORKDIR /app
COPY package*.json ./
RUN npm install --production
COPY . .
RUN npm run build

FROM alpine:3.18
WORKDIR /app
COPY --from=builder /app/dist/ ./dist/
CMD ["node", "dist/index.js"]`;

  // Write sample to temp file for demo
  const tempFile = path.join(__dirname, 'temp_demo.Dockerfile');
  fs.writeFileSync(tempFile, sampleDockerfile);
  
  try {
    console.log('=== ShipCheck Layer Size Calculator ===\n');
    
    const result: SizeSummary = analyzeDockerfile(sampleDockerfile);
    
    console.log(`Total Image Size Estimate: ${result.totalSizeMB.toFixed(2)} MB\n`);
    console.log(`Breakdown:`);
    console.log(`  - Base Image:     ${result.baseImageMB.toFixed(2)} MB`);
    console.log(`  - Runtime Commands: ${result.runtimeCommandsMB.toFixed(2)} MB`);
    console.log(`  - File Copies:     ${result.fileCopiesMB.toFixed(2)} MB`);
    console.log(`\nLayer Details:`);
    
    for (const layer of result.layers) {
      const padding = ' '.repeat(40 - layer.instruction.length);
      console.log(`  [${layer.lineNumber}] ${layer.instruction}${padding} ~${layer.estimatedSizeMB.toFixed(2)} MB`);
      if (layer.description) {
        console.log(`    └─ ${layer.description}`);
      }
    }
    
    // Cleanup temp file
    fs.unlinkSync(tempFile);
  } catch (error: any) {
    console.error('Error:', error.message);
    process.exit(1);
  }
}

export default analyzeDockerfile;